CRM/Membershipextrasimporterapi/EntityImporter/LineItem.php

```diff
@@ -0,0 +1,303 @@
+      'line_item_label' => $contributionSqlParams[6][0],
+      'quantity' => $contributionSqlParams[7][0],
+      'unit_price' => $contributionSqlParams[8][0],
+      'line_total' => $contributionSqlParams[9][0],
+      'financial_type_id' => $contributionSqlParams[10][0],
+    ];
+  }
+
+  private function setPriceFieldValueDetails() {
+    if (!empty($this->rowData['line_item_price_field_id']) && !empty($this->rowData['line_item_price_field_value_id'])) {
+      $this->priceFieldValue = $this->getPriceFieldValueDetailsById($this->rowData['line_item_price_field_value_id']);
+      return;
+    }
+
+    if ($this->entityTable == 'civicrm_membership') {
+      $this->priceFieldValue = $this->getMembershipPriceFieldValueDetails();
+    }
+    else {
+      $defaultContributionAmountPriceFieldValueId = 1;
+      $this->priceFieldValue = $this->getPriceFieldValueDetailsById($defaultContributionAmountPriceFieldValueId);
+    }
+  }
+
+  private function getPriceFieldValueDetailsById($priceFieldValueId) {
+    $dao = CRM_Core_DAO::executeQuery("SELECT * FROM civicrm_price_field_value WHERE id = {$priceFieldValueId}");
+    $dao->fetch();
+    return $dao->toArray();
+  }
+
+  private function getMembershipPriceFieldValueDetails() {
+    $membershipTypeId = $this->membership['membership_type_id'];
+    $dao = CRM_Core_DAO::executeQuery("SELECT * FROM civicrm_price_field_value WHERE membership_type_id = {$membershipTypeId} 
+                                       ORDER BY id ASC LIMIT 1");
+    $dao->fetch();
+    return $dao->toArray();
+  }
+
+  private function getLineItemLabel() {
+    if (!empty($this->rowData['line_item_label'])) {
+      return $this->rowData['line_item_label'];
+    }
+
+    if ($this->entityTable == 'civicrm_membership') {
+      return $this->priceFieldValue['label'];
+    }
+    else {
+      return 'Donation';
+    }
+  }
+
+  private function getQuantity() {
+    if (!empty($this->rowData['line_item_quantity'])) {
+      return $this->rowData['line_item_quantity'];
+    }
+
+    return 1;
+  }
+
+  private function getUnitPrice() {
+    return $this->rowData['line_item_unit_price'];
+  }
+
+  private function getLineTotal($unitPrice, $quantity) {
+    return MoneyUtilities::roundToCurrencyPrecision($unitPrice * $quantity);
+  }
+
+  private function getFinancialTypeId() {
+    if (!isset($this->cachedValues['financial_types'])) {
+      $sqlQuery = "SELECT id, name FROM civicrm_financial_type";
+      $result = CRM_Core_DAO::executeQuery($sqlQuery);
+      while ($result->fetch()) {
+        $this->cachedValues['financial_types'][$result->name] = $result->id;
+      }
+    }
+
+    if (!empty($this->cachedValues['financial_types'][$this->rowData['line_item_financial_type']])) {
+      return $this->cachedValues['financial_types'][$this->rowData['line_item_financial_type']];
+    }
+
+    throw new CRM_Membershipextrasimporterapi_Exception_InvalidLineItemException('Invalid line item "Financial Type"', 200);
```

Comment:
Should we add localisation function ts() to the error message as well?

Reply:
Even that we show these during the import if there is any error, these exception messages are kinda technical, and I am not sure if localizing them make much sense, so will keep them as is.